src/catalog/catalog_accessor.cpp

```diff
@@ -178,6 +178,10 @@ common::ManagedPointer<execution::udf::UDFContext> CatalogAccessor::GetUDFContex
   return dbc_->GetUDFContext(txn_, proc_oid);
 }
 
+std::unique_ptr<optimizer::TableStats> CatalogAccessor::GetTableStats(table_oid_t table_id) {
```

Comment:
I do not understand why this is a unique pointer? Does the stats object is deleted as soon as it is fetched? Why isn't it a managed pointer owned by someone else? We should not create this object every time while doing query optimisation for every table.

Reply:
Thanks for the comments! This is definitely a good point in terms of efficiency, but I am not sure I can think of a significantly better solution.

In the existing interface, the StatsStorage object owns an unordered_map of containing TableStats objects, and thus looking up the statistics for a table is as simple as returning a `common::ManagedPointer<TableStats>`. However, since TableStats are only kept in memory, they are not persisted and will need to be recomputed each time after the database is restarted.

Instead, this commit stores the statistics data in the pg_statistics table. However, in order to work with the cost model code (which currently expects a TableStats object), this function will need to create a copy of the table data to fit into the TableStats object.

You're right that this will lead to more allocations than the existing StatsStorage API. However, I am not sure that there are better alternatives:

* The returned TableStats storage objects could be cached at some layer. However, this means that the ANALYZE command would need to invalidate that cache, which leads to tricky design issues, including lifetime issues that might require reference counting to resolve. For example, what if ANALYZE is running at the same time that a query is being executed on the same table — how can we ensure the statistics are read correctly? Solving these issues would lead to essentially re-implementing the storage layer: with the current approach, the guarantees of MVCC prevent us from having to worry about this.
* The cost model code could be changed to use a different API that requires less allocations. I am not familiar enough to say whether this is the case or not, but I am not sure it will be possible to make an API with significantly better performance characteristics in terms of allocations.

Given these options, I think it wouldn't be worth making things more complicated than they already are. Please let me know what you think, or if there's any other way the allocation cost can be reduced!

Reply:
(And in fact, I think some of those lifetime issues could possibly exist in the existing StatsStorage API, if the same TableStats object is being updated by one thread at the same time that it is being read by another...?)